runtime/cudaq/builder/kernels.cpp

```diff
@@ -7,181 +7,99 @@
  ******************************************************************************/
 
 #include "kernels.h"
-#include "common/EigenDense.h"
 
 namespace cudaq::details {
 
-std::vector<std::string> grayCode(std::size_t rank) {
-  std::function<void(std::vector<std::string> &, std::size_t)> grayCodeRecurse;
-  grayCodeRecurse = [&grayCodeRecurse](std::vector<std::string> &g,
-                                       std::size_t rank) {
-    auto k = g.size();
-    if (rank <= 0)
-      return;
-
-    for (int i = k - 1; i >= 0; --i) {
-      auto c = "1" + g[i];
-      g.push_back(c);
-    }
-    for (int i = k - 1; i >= 0; --i) {
-      g[i] = "0" + g[i];
-    }
-
-    grayCodeRecurse(g, rank - 1);
-  };
-
-  std::vector<std::string> g{"0", "1"};
-  grayCodeRecurse(g, rank - 1);
-
-  return g;
+std::vector<std::size_t> grayCode(std::size_t numBits) {
+  std::vector<std::size_t> result(1u << numBits);
+  for (std::size_t i = 0; i < (1ULL << numBits); ++i)
+    result[i] = ((i >> 1) ^ i);
+  return result;
 }
 
-std::vector<std::size_t> getControlIndices(std::size_t grayRank) {
-  auto code = grayCode(grayRank);
-  std::vector<std::size_t> ctrlIds;
-  for (std::size_t i = 0; i < code.size(); i++) {
-    auto a = std::stoi(code[i], nullptr, 2);
-    auto b = std::stoi(code[(i + 1) % code.size()], nullptr, 2);
-    auto c = a ^ b % code.size();
-    ctrlIds.emplace_back(std::log2(c));
+std::vector<std::size_t> getControlIndices(std::size_t numControl) {
+  auto code = grayCode(numControl);
+  std::vector<std::size_t> indices;
+  for (auto i = 0u; i < code.size(); ++i) {
+    // The position of the control in the lth CNOT gate is set to match
+    // the position where the lth and (l + 1)th bit strings g[l] and g[l+1] of
+    // the binary reflected Gray code differ.
+    auto position = std::log2(code[i] ^ code[(i + 1) % code.size()]);
+    // The bit position will be given in littlen endian, i.e., bit 0 is the
+    // rightmost bit. In CUDA Quantum, we use big endian, thus we need to
```

Comment:
Doesn't "bit 0 is the rightmost bit" mean to big endian? I.e. bit positions `76543210` refer to a big endian number, and `0` is the rightmost bit.

Also "littlen" --> "little".

Reply:
I will improve the comment to talk about most significant bit vs least significant bit since big endian vs little endian is more often used in the context of byte ordering. 

> Doesn't "bit 0 is the rightmost bit" mean to big endian?

No. It's the other way around. Take the following example of representing `1` with three bits:
```
msb -v
     001
       ^- lsb

                        0 0 1
little endian indices:  2 1 0  (lsb in the smallest index, aka lsb0)
big endian indices:     0 1 2  (msb in the smallest index, aka msb0)
```

Reply:
Hmm, that's not how I've used it before, but as you say, it is normally referring to byte ordering. In byte ordering, this is big endian: B3:B2:B1:B0, where B3 is the most significant byte and B0 is the least significant byte. And correspondingly, little endian is B0:B1:B2:B3. In order to keep neighboring bits side by side (e.g. the most significant bit of B1 of and the least significant bit of B2), that would make the little endian bits be 0 (LSB), 1, ... 31 (MSB).

Reply:
Hm, aren't we talking about the same thing? It is just that you are placing the bytes in a container that is indexed from left to right. If I changed my bit explanation to bytes:
```
msB -v
     B3 B2 B1 B0
               ^- lsB

                        B3 B2 B1 B0 
little endian indices:  3  2  1  0  (lsB in the smallest index)
big endian indices:     0  1  2  3  (msB in the smallest index)
```
 
If I list this: `big endian: B3 B2 B1 B0`, `little endian: B0 B1 B2 B3`

Reply:
Here is a link with generally accepted [definitions](https://stackoverflow.com/questions/16803397/can-endianness-refer-to-bits-order-in-a-byte).

What we're talking about, I think, is the serialization of bits. Conventionally, binary numbers are written from right-to-left as significance increases. The "IBM" notation uses a decreasing index, inversely to the bit significance. That's "convenient" if one has a predetermined number of bits and wants to print from index 0 to $n-1$, in that order, and see a conventional binary number as output. Otherwise the bits are numbered corresponding to increasing significance of the bits. Bit 0 is thus the least significant, and so, up to the most significant bit being the highest index, e.g., 63. It is just as easy to get the same conventional binary number as output; just use a decrementing loop.

PS: I know we all know this stuff, but the overloading of byte-addressable terms to this conversation is causing some confusion. ;)

Reply:
Yes, I think we're on the same page, but I am still concerned by the "bit 0" comment in the code because I think "bit 0" is nearly always referring to the least significant bit (by definition of it representing the weight for 2^`0`), so I still think that "bit 0 is the rightmost bit" is only true for big endian representations (which place LSBs at the right). Your diagram has _index_ 0 as the rightmost _index_ in `little endian indices`, and I agree with that, but that's different than saying "bit 0 is the rightmost bit".

Reply:
> so I still think that "bit 0 is the rightmost bit" is only true for big endian representations (which place LSBs at the right). Your diagram has _index_ 0 as the rightmost _index_ in `little endian indices`, and I agree with that, but that's different than saying "bit 0 is the rightmost bit".

It depends on how you index you container right. We all agree that endianness has to do in which index the most (least) significant bit is stored. In this case, I did mean least significant bit when I said "bit 0". In programing languages it might uncommon to see a container indexed from left to right, but in HDL is a common thing:

```vhdl
signal little_endian : std_logic_vector(3 downto 0);
-- If I put the number 1 (001) in this vector, what do I get when I do little_endian[0]?
```

We get `1` right? The least significant bit (bit 0) is the rightmost position. Basically the bitstring representing the number stays the same, we just decided to index the bitstring "3 downto 0". (For big endian we would do "0 to 3", and then index 2 would return `1`. See [here](https://en.wikipedia.org/wiki/Endianness#Logic_design))

That being said, in CUDA Quantum when we see the bistring `001` is doesn't mean the number 1, it means 4. Thus we consider the least significant bit (bit 0) being on the left. This is the difference, so I will improved the comment to reflect that.

Reply:
I fear that left/right, least/most-significant, and little/big-endian aren't going to get us out of the box in explaining what's happening/expected. This conversation more or less shows that, eh?

We need something more along the lines of each qubit is assigned a logical(?) unique id from $0$ to $N-1$. Then if we want to say how these are displayed (and read?), we can say the qubits are printed from left to right, ids number $0$ through $N-1$, in a bitstring of length $N$. (Which is admittedly contrary to normal mathematical convention, and either a feature or a bug.)